### PR TITLE
fixes #1364 Process all eligible IndexStreams in Union.advanceQueue()

### LINKED ISSUE #1364: Range stream misses shards edge case
query && ((function && (SHARDS_AND_DAYS = 'X,Y,Z')) || (function && (SHARDS_AND_DAYS = 'X,Y,Z')))
where X,Y, and Z are days
All of the actually documents are in Z, but it only returns ranges for the first shard in Z.

Interestingly the following variants succeed in finding all of the ranges:
query && ((function && (SHARDS_AND_DAYS = 'Z')) || (function && (SHARDS_AND_DAYS = 'Z')))
query && ((function && (SHARDS_AND_DAYS = 'X,Z')) || (function && (SHARDS_AND_DAYS = 'X,Z')))

### PR DESCRIPTION
**Pull request by FineAndDandy**

### CODE REVIEW

File: warehouse/query-core/src/main/java/datawave/query/index/lookup/Union.java [hunk 1]
```diff
@@ -198,13 +211,21 @@ public boolean hasNext() {
             pointers = pointers.union(itr.peek().second(), Lists.newArrayList(delayedNodes.getNodes()));
             itr.next();
             if (itr.hasNext()) {
-                children.add(itr);
+                // add this to the processed list so other iterators have a chance to be first even if this one comes back earlier in the stack
+                processedChildren.add(itr);
+                // children.add(itr);
```
Comment (apmoriarty) [inline warehouse/query-core/src/main/java/datawave/query/index/lookup/Union.java]: remove this comment

**Review by apmoriarty — DISMISSED** [review]

**Review by alerman — APPROVED** [review]

File: warehouse/query-core/src/main/java/datawave/query/index/lookup/Union.java [hunk 2]
```diff
@@ -198,13 +211,20 @@ public boolean hasNext() {
             pointers = pointers.union(itr.peek().second(), Lists.newArrayList(delayedNodes.getNodes()));
             itr.next();
             if (itr.hasNext()) {
-                children.add(itr);
+                // add this to the processed list so other iterators have a chance to be first even if this one comes back earlier in the stack
+                processedChildren.add(itr);
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/index/lookup/Union.java:215]: Do we always need to add it back in or only if it is a day?

File: warehouse/query-core/src/main/java/datawave/query/index/lookup/Union.java [hunk 3]
```diff
@@ -198,13 +211,20 @@ public boolean hasNext() {
...
             } else {
                 if (log.isTraceEnabled()) {
                     log.trace("IndexStream exhausted for " + itr.getContextDebug());
                 }
             }
             childrenAdded = true;
+            
+            // repopulate children with all processed children
+            if (children.isEmpty() && !processedChildren.isEmpty()) {
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/index/lookup/Union.java:224]: Do the children need to be empty to add them in here?  And then if that is the case perhaps the code from 198-205 is not needed?

**Review by ivakegg — CHANGES_REQUESTED** [review]

File: warehouse/query-core/src/main/java/datawave/query/index/lookup/Union.java (hunk 2 above)
Comment (ivakegg) [reply warehouse/query-core/src/main/java/datawave/query/index/lookup/Union.java:215]: ah I see.... just adding it into a new list.

File: warehouse/query-core/src/main/java/datawave/query/index/lookup/Union.java (hunk 3 above)
Comment (ivakegg) [reply warehouse/query-core/src/main/java/datawave/query/index/lookup/Union.java:224]: Ah I see.... need to allow children to get empty first.

**Review by ivakegg — APPROVED** [review]

**Pull request merged**